substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/sampler/SamplerBufferList.java

```diff
@@ -0,0 +1,94 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.sampler;
+
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.word.WordFactory;
+
+import com.oracle.svm.core.thread.VMOperation;
+import com.oracle.svm.core.jfr.BufferNodeAccess;
+import com.oracle.svm.core.jfr.BufferNode;
+import com.oracle.svm.core.jfr.BufferList;
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.jfr.sampler.AbstractJfrExecutionSampler;
+
+/**
+ * Nodes should only be removed from this list by
+ * {@link SamplerBuffersAccess#processActiveBuffers()}. Nodes are marked for removal if their buffer
+ * field is null. This means that the buffer has been put on the full buffer queue because it is
+ * full or the owning thread has exited.
+ *
+ * This class also contains some checks to ensure the BufferList lock won't be held while sampling
+ * is enabled. Otherwise, deadlock can occur with a single thread due to recursive locking. This is
+ * because the sampler code may add a new node to the {@link SamplerBufferList} when the thread's
+ * current {@link SamplerBuffer} reaches capacity.
+ */
+public class SamplerBufferList extends BufferList {
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public SamplerBufferList() {
+    }
+
+    /**
+     * This is called after {@link SamplerBufferPool#teardown()}, so all buffers should already be
+     * freed. All nodes in the list can be freed.
+     */
+    public void teardown() {
+        assert VMOperation.isInProgressAtSafepoint();
+
+        BufferNode node = head;
+        while (node.isNonNull()) {
+            assert BufferNodeAccess.getSamplerBuffer(node).isNull();
+            BufferNode next = node.getNext();
+            BufferNodeAccess.free(node);
+            node = next;
+        }
+        head = WordFactory.nullPointer();
+    }
+
+    @Override
+    @Uninterruptible(reason = "Locking without transition requires that the whole critical section is uninterruptible.")
+    public BufferNode getHead() {
+        assert !AbstractJfrExecutionSampler.isExecutionSamplingAllowedInCurrentThread();
+        return super.getHead();
+    }
+
+    @Override
+    @Uninterruptible(reason = "Locking without transition requires that the whole critical section is uninterruptible.")
+    public BufferNode addNode(com.oracle.svm.core.jfr.Buffer buffer) {
+        assert !AbstractJfrExecutionSampler.isExecutionSamplingAllowedInCurrentThread();
```

Comment:
Thanks for doing the changes.

In the last review, I missed that this is called from the signal handler. This is not allowed because in the signal handler, we can't allocate *any* memory (i.e., not even C memory).

Reply:
Oh ok I see. I've changed things now so that the `BufferNode` allocations are done by `SamplerBufferPool` at the same time `SamplerBuffer` allocations are done.  Now, buffers acquired from the "available" stack already have corresponding nodes, which are ready to be linked onto `SamplerBufferList` when the buffers become active. When full buffers are released after being serialized, new `BufferNode`s are allocated for them before they go back into circulation on the available list.